---
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    lib_dir: libs
css: styles.css
params: 
    title: "GRIN Epilepsy Trio Analysis!"
---
---
title: `r params$title`
---
```{r libs, echo=FALSE, message=FALSE, cache=FALSE, warning=FALSE, results='hide', fig.path='{{SLINK}}/'}
library(knitr)
library(DT)
library(plyr)
library(dplyr)
library(GenomicRanges)
options(digits = 2)
options(width=1200)
```

```{r functions, echo=FALSE, message=FALSE, cache=FALSE, warning=FALSE, results='hide', fig.path='{{SLINK}}/'}
## http://www.r-bloggers.com/identifying-records-in-data-frame-a-that-are-not-contained-in-data-frame-b-%E2%80%93-a-comparison/
thisnotinthat <- function(x.1,x.2,...){
  x.1p <- do.call("paste", x.1)
  x.2p <- do.call("paste", x.2)
  x.1[! x.1p %in% x.2p, ]
}

viewDataTable <- function(dat){

  cols2hide <- grep('Description|Variants|Perc|chr|^pos$|rsid|^ref$|^alt$|^gene$|
                    impact_severity|rvis_score|depths|gt_quals|family_genotypes|is_exonic|is_splicing',
                    colnames(dat), invert = TRUE)

  DT::datatable(dat,
                extensions = c('Buttons'),
                selection = "single",
                filter = "bottom",
                options = list(
                  columnDefs = list(list(visible = FALSE, targets = cols2hide)),
                  dom = 'Bfrtip',
                  buttons = list('colvis','pageLength', 'copy','print',
                                 list(extend = "collection",
                                      buttons = c('csv', 'excel', 'pdf'),
                                      text = 'Download'
                                 )),
                  searchHighlight = TRUE,
                  lengthMenu = list(c(5, 10, 15, 20, 25, -1), c('5', '10', '15', '20', '25', 'All')),
                  initComplete = DT::JS("function(settings, json) {",
                                        "$(this.api().table().header()).css({'background-color': '#003366', 'color': '#fff'});",
                                        "}"),
                  scrollX = TRUE
                ))
}
```

## Filters
<blockquote>
We used the following filters to reduce the number of reported false positives:

1. Spurious inherited variants can often appear due to insufficient sequence coverage so the **mininum sequence depth** was set to **15**.

2. We are interested in mutations that are likely to have functional consequence so only variants with **Medium or High impact severity** were used.
</blockquote>


## Compound Heterozygotes

Many autosomal recessive disorders are caused by compound heterozygotes. Unlike canonical recessive sites where the same recessive allele is inherited from both parents at the _same_ site in the gene, compound heterozygotes occur when the individualâ€™s phenotype is caused by two heterozygous recessive alleles at _different_ sites in a particular gene.
<blockquote>
Genotype Requirements:
  
1. All affected individuals must be heterozygous at both sites.
2. No unaffected can be homozygous alterate at either site.
3. Neither parent of an affected sample can be homozygous reference at both sites.
4. If any unphased-unaffected is het at both sites, the site will be give lower priority
5. No phased-unaffected can be heterozygous at both sites.
6. Remove candidates where an affected from the same family does NOT share the same het pair.
7. Sites are automatically phased by transmission when parents are present in order to remove false positive candidates.
</blockquote>

<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Variant summary for Compound Heterozygotes</h3>
</div>
<div class="panel-body">

This represents a brief summary of the Compound Heterozygotes.

```{r cmpdhet_summary, warning = FALSE, echo=FALSE, message=FALSE, cache=FALSE,  fig.path='{{SLINK}}/'}

if(length(cmpdhet) == 1 && cmpdhet == 0)
{
  varcount <- 0
  showChunk <- FALSE
} else {
  showChunk <- TRUE
  dt <- plyr::count(cmpdhet , 'impact_so')
  viewDataTable(dat = data.frame(Description = dt$impact_so, 
                                 Variants = dt$freq, 
                                 Perc = round(prop.table(dt$freq)*100,2)))
}
```

```{r cmpdhet_rvis, warnings = FALSE, eval=showChunk, echo=FALSE, message=FALSE, cache=FALSE,fig.path='{{SLINK}}/'}
cmpdhet <- merge(cmpdhet, rvis, by.x = "gene", by.y = "GENE", all.x = T, sort = F)
colnames(cmpdhet)[grep('RVIS',colnames(cmpdhet))] <- "rvis_score"
```

```{r cmpdhet_pp, warnings = FALSE, eval=showChunk, echo=FALSE, message=FALSE, cache=FALSE,fig.path='{{SLINK}}/'}
cmpdhet$start <- ifelse(cmpdhet$end-cmpdhet$start==1, cmpdhet$end, paste0(cmpdhet$start,'-',cmpdhet$end))
colnames(cmpdhet)[grep('chr|start|vcf_id', colnames(cmpdhet))] <- c('chr','pos','rsid')
cmpdhet$end <- NULL
cmpdhet <- cmpdhet[,c(2:16,1,17:ncol(cmpdhet))]
```

```{r cmpdhet_filtered_variants, warning = FALSE, eval=showChunk, echo=FALSE, message=FALSE, cache=FALSE,  fig.path='{{SLINK}}/'}
varcount <- length(unique(cmpdhet$variant_id))
```

**There are `r varcount` compound heterozygotes.**
</div>
</div>

#### Triage comphet
<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Triage 1: Compound Heterozygotes in recurrent genes</h3>
</div>
<div class="panel-body">
These are variants found in the genes associated with recurrent mutations in epilepsy.

```{r cmpdhet_recurrent, eval=showChunk, warning = FALSE, echo=FALSE, message=FALSE, cache=FALSE,  fig.path='{{SLINK}}/'}
cmpdhet %>% dplyr::filter(recurrent_genes != "None") -> in_recurrent
varcount <- length(unique(in_recurrent$variant_id))
if(varcount)
{
  viewDataTable(dat = in_recurrent)
}
```
**There are `r varcount` remaining variants found in recurrent genes.**
</div>
</div>


<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Triage 2: Compound Heterozygotes in literature genes with low tolerance</h3>
</div>
<div class="panel-body">
These are the remaining variants cited in the genes cited in the existing literature. 
Many of the literature-cited genes are likely tolerant, so this list is further filtered to only genes with an RVIS "ALL_0.1." value of <= 0.

```{r cmpdhet_literature, eval=showChunk, warning = FALSE, echo=FALSE, message=FALSE, cache=FALSE,  fig.path='{{SLINK}}/'}
cmpdhet %>% dplyr::filter(literature_genes != "None") %>% dplyr::filter(as.numeric(rvis_score)<=0) %>% thisnotinthat(in_recurrent) -> in_literature
varcount <- length(unique(in_literature$variant_id))
if(varcount)
{
  viewDataTable(dat = in_literature)
}
```
**There are `r varcount` remaining variants found in literature genes.**
</div>
</div>


<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Triage 3: Comp het coding variants not in the above</h3>
</div>
<div class="panel-body">
These are remaining variants that are located in the coding sequence of at least one transcript.

```{r cmpdhet_coding, eval=showChunk, warning = FALSE, echo=FALSE, message=FALSE, cache=FALSE,  fig.path='{{SLINK}}/'}
cmpdhet %>% dplyr::filter(is_coding==1) %>% thisnotinthat(in_literature) -> codingonly
varcount <- length(unique(codingonly$variant_id))
if(varcount)
{
  viewDataTable(dat = codingonly)
}
```
**There are `r varcount` remaining variants found in coding genes**
</div>
</div>



<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Other Compound Heterozygotes</div>
<div class="panel-body">
These are the remaining compound heterozygotes

```{r cmpdhet_others, eval=showChunk, warning = FALSE, echo=FALSE, message=FALSE, cache=FALSE,  fig.path='{{SLINK}}/'}
others <- thisnotinthat(cmpdhet, in_literature)
others <- thisnotinthat(others, codingonly)

varcount <- length(unique(others$variant_id))
if(varcount)
{
  viewDataTable(dat = others)
}
```
**There are `r varcount` remaining variants**
</div>
</div>
