---
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    lib_dir: libs
css: styles.css
params: 
    title: "GRIN Epilepsy Trio Analysis!"
---
---
title: `r params$title`
---
```{r libs, echo=FALSE, message=FALSE, cache=FALSE, warning=FALSE, results='hide', fig.path='{{SLINK}}/'}
library(knitr)
library(DT)
library(plyr)
library(dplyr)
library(GenomicRanges)
library(stringr)
library(reshape2)
options(digits = 2)
options(width=1200)
```

```{r commit, echo=FALSE, message=FALSE, cache=FALSE, warning=FALSE, results='hide', fig.path='{{SLINK}}/'}
commit <- system("git rev-parse --verify HEAD",intern=TRUE)
```

## Filters
<blockquote>
We used the following filters to reduce the number of reported false positives:

1. Spurious inherited variants can often appear due to insufficient sequence coverage so the **mininum sequence depth** was set to **15**.
2. We are interested in mutations that are likely to have functional consequence so only variants with **Medium or High impact severity** were used.
3. **Minimum genotyping quality** per sample in the trio was set to **20**.
4. GL contigs were not used.
</blockquote>

## De novo analysis
<blockquote>
Genotype Requirements:
  
1. all affecteds must be het
2. all unaffected must be homref or homalt
3. at least 1 affected kid must have unaffected parents
4. if an affected has affected parents, itâ€™s not de_novo
5. all affected kids must have unaffected (or no) parents
6. warning if none of the affected samples have parents.
</blockquote>

<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Variant summary</h3>
</div>
<div class="panel-body">

This represents a brief summary of variants.

```{r var_summary, results='asis', warning=FALSE, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}

if(length(var.df) == 1 && var.df == 0)
{
  varcount <- 0
  showChunk <- FALSE
} else {
  showChunk <- TRUE
  dt <- plyr::count(var.df , 'impact_so')
  viewDataTable(dat = data.frame(Description = dt$impact_so, 
                                 Variants = dt$freq, 
                                 Perc = round(prop.table(dt$freq)*100,2)))
}
```

```{r var_igv, warnings=FALSE, eval=showChunk, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
var.df$variant_id <- paste("<a href=\"", paste(paste("../screenshots", paste(familyname, model, var.df$variant_id, 'png', sep="."), sep="/")), "\">", var.df$variant_id, "</a>", sep="")
```

```{r var_rvis, warnings=FALSE, eval=showChunk, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
var.df$orig_chrom <- var.df$chrom
var.df$chrom <- sub('^chr','',var.df$chrom)
var.df <- merge(var.df, rvis, by.x = "gene", by.y = "GENE", all.x = T, sort = F)
```

```{r var_pp, warnings=FALSE, eval=showChunk, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
denovogear <- denovogear[-grep('^GL',denovogear$chr),]
denovogear$chr <- sub('^chr','',denovogear$chr)
gr1 <- with(var.df, GRanges(chrom, IRanges(start = start, end = end))) 
gr2 <- with(denovogear, GRanges(chr, IRanges(start = pos, end = pos)))
type1 <- findOverlaps(query = gr1, subject = gr2, type = 'any')
var.df$denovo_prob <- NA
var.df[queryHits(type1),'denovo_prob'] <- denovogear[subjectHits(type1),'pp_dnm']
var.df <- var.df[order(var.df$denovo_prob, decreasing = TRUE),]
```

```{r var_formatcols, warnings=FALSE, eval=showChunk, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
var.df$start <- ifelse(var.df$end-var.df$start==1, var.df$end, paste0(var.df$start,'-',var.df$end))
colnames(var.df)[grep('^chrom$|start|vcf_id', colnames(var.df))] <- c('chr','pos','rsid')
var.df$end <- NULL
var.df <- var.df[,c(2:16,1,17:ncol(var.df))]

chr <- c(1:22,'X','Y','MT')
var.df$chr <- factor(var.df$chr, levels = chr)
var.df <- var.df[order(var.df$chr),]
var.df$exac_ac_all[var.df$exac_ac_all=="None"] <- "-1"
var.df$exac_ac_all <- as.numeric(var.df$exac_ac_all)
```

```{r var_filtered_variants, warning=FALSE, eval=showChunk, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
varcount <- length(unique(var.df$variant_id))
```

**There are `r varcount` variants.**
</div>
</div>

<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">All variants</h3>
</div>
<div class="panel-body">
These are all variants

```{r var_all, eval=showChunk, warning=FALSE, echo=FALSE, message=FALSE, cache=FALSE,  fig.path='{{SLINK}}/'}
var.df.all <- var.df

varcount <- length(unique(var.df.all$variant_id))
if(varcount)
{
  viewDataTable(dat = var.df.all)
}
```
**There are a total of `r varcount` variants**
</div>
</div>

<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Variants for Ingo Helbig</h3>
</div>
<div class="panel-body">
These are all variants that satisfy the following criteria:

1. Variant function must be frameshift/missense/unknown.
2. Exac Adjusted allele count must be 0 or unknown.
3. Denovogear posterior probability must be >=20%.
4. Should be exonic and splicing.

```{r var_filtered, eval=showChunk, warning=FALSE, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
var.df.sub <- var.df
var.df.sub <- var.df.sub[which(var.df.sub$is_exonic==1 & var.df.sub$is_splicing==1),]
var.df.sub <- var.df.sub[which(var.df.sub$functiongvs == "frameshift" | var.df.sub$functiongvs == "missense" | var.df.sub$functiongvs == "None"),]
var.df.sub <- var.df.sub[which(var.df.sub$exac_ac_all == "-1" | var.df.sub$exac_ac_all == "0"),]
var.df.sub <- var.df.sub[which(var.df.sub$denovo_prob >= 0.20),]

varcount <- length(unique(var.df.sub$variant_id))
if(varcount)
{
  viewDataTable(dat = var.df.sub)
}
```
**There are `r varcount` filtered variants**
</div>
</div>

<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Variants for Ingo Helbig (Enhanced)</h3>
</div>
<div class="panel-body">
These are all variants that satisfy the following criteria:

1. Allelic ratio for proband between 0.20-0.80
2. Allelic ratio for mom and dad: 0

```{r var_filtered_enhancement, eval=showChunk, warning=FALSE, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
var.df.sub <- var.df
alt_depths <- colsplit(gsub("\\s+"," ", str_trim(gsub('[[]|[]]|[.]',' ',var.df.sub$gt_alt_depths)))," ",c('mom','dad','kid'))
gt_depths <- colsplit(gsub("\\s+"," ", str_trim(gsub('[[]|[]]|[.]',' ',var.df.sub$gt_depths)))," ",c('mom','dad','kid'))
dp <- data.frame(mom = alt_depths[,1]/gt_depths[,1], dad = alt_depths[,2]/gt_depths[,2], proband = alt_depths[,3]/gt_depths[,3])
var.df.sub <- var.df.sub[which(dp$proband>0.20 & dp$proband<0.80 & dp$mom==0 & dp$dad==0),]
varcount <- length(unique(var.df.sub$variant_id))
if(varcount)
{
  viewDataTable(dat = var.df.sub)
}
```
**There are `r varcount` filtered variants**
</div>
</div>


<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Meta</h3>
</div>
<div style="color:#003366;", class="panel-body">
Git commit: **`r commit`**
</div>
</div>
