---
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    lib_dir: libs
css: styles.css
params: 
    title: "GRIN Epilepsy Trio Analysis!"
---
---
title: `r params$title`
---
```{r libs, echo=FALSE, message=FALSE, cache=FALSE, warning=FALSE, results='hide', fig.path='{{SLINK}}/'}
library(knitr)
library(DT)
library(plyr)
library(dplyr)
library(GenomicRanges)
library(stringr)
library(reshape2)
options(digits = 2)
options(width=1200)
```

```{r commit, echo=FALSE, message=FALSE, cache=FALSE, warning=FALSE, results='hide', fig.path='{{SLINK}}/'}
commit <- system("git rev-parse --verify HEAD",intern=TRUE)
```

## Filters
<blockquote>
We used the following filters to reduce the number of reported false positives:

1. Spurious inherited variants can often appear due to insufficient sequence coverage so the **mininum sequence depth** was set to **15**.
2. We are interested in mutations that are likely to have functional consequence so only variants with **Medium or High impact severity** were used.
3. **Minimum genotyping quality** per sample in the trio was set to **20**.
4. GL contigs were not used.
5. Genotype Quality > 80 for all samples
6. ExAC frequency of 0 (or unknown)
7. RVIS genic intolerance percentile < 90%
</blockquote>

## De novo analysis
<blockquote>
Genotype Requirements:
  
1. all affecteds must be het
2. all unaffected must be homref
3. at least 1 affected kid must have unaffected parents
4. if an affected has affected parents, itâ€™s not de_novo
5. all affected kids must have unaffected (or no) parents
6. warning if none of the affected samples have parents.
</blockquote>

```{r dnm, warnings=FALSE, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
var.df<-execdnm.df
```

```{r var_filtered_variants, warning=FALSE, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
varcount <- length(unique(var.df$variant_id))
if(length(var.df) == 1 && var.df == 0)
{
  varcount <- 0
  showChunk <- FALSE
} else {
  showChunk <- TRUE
}
```

```{r var_pp, warnings=FALSE, eval=showChunk, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
denovogear <- denovogear[-grep('^GL',denovogear$chr),]
denovogear$chr <- sub('^chr','',denovogear$chr)

gr1 <- with(var.df, GRanges(chrom, IRanges(start = start, end = end))) 
gr2 <- with(denovogear, GRanges(chr, IRanges(start = pos, end = pos)))

type1 <- findOverlaps(query = gr1, subject = gr2, type = 'any')
var.df$denovo_prob <- NA
var.df[queryHits(type1),'denovo_prob'] <- denovogear[subjectHits(type1),'pp_dnm']
var.df <- var.df[order(var.df$denovo_prob, decreasing = TRUE),]
```

```{r var_format, warnings=FALSE, eval=showChunk, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
var.df$variant_id <- paste("<a href=\"", paste(paste("../screenshots", paste(familyname, model, var.df$variant_id, 'png', sep="."), sep="/")), "\">", var.df$variant_id, "</a>", sep="")
var.df$orig_chrom <- var.df$chrom
var.df$chrom <- sub('^chr','',var.df$chrom)
var.df$start <- ifelse(var.df$end-var.df$start==1, var.df$end, paste0(var.df$start,'-',var.df$end))
colnames(var.df)[grep('^chrom$|start|vcf_id', colnames(var.df))] <- c('chr','pos','rsid')
var.df$end <- NULL
var.df <- var.df[,c(2:16,1,17:ncol(var.df))]

chr <- c(1:22,'X','Y','MT')
var.df$chr <- factor(var.df$chr, levels = chr)
var.df <- var.df[order(var.df$chr),]
```


<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">De novo</h3>
</div>
<div class="panel-body">
Variants matching following gemini query:
```
gemini query -q "select * from variants where 1 and chrom NOT LIKE '%GL%' and (exac_an_all < 0.1 or exac_an_all is null) and (rvis_pct < 90) \
and (impact != 'intron_variant' and impact != '3_prime_UTR_variant' and impact != '5_prime_UTR_variant' )" \
--gt-filter "(gt_types.{mom} == HOM_REF) and (gt_types.{dad} == HOM_REF) \
and (gt_types.{proband} == HET) and (gt_depths).(*).(>=15).(all) and (gt_quals).(*).(>=80).(all)" \
--header {db} > {out}
```

```{r var_dnm, eval=showChunk, warning=FALSE, echo=FALSE, message=FALSE, cache=FALSE,  fig.path='{{SLINK}}/'}
var.df.all <- var.df

varcount <- length(unique(var.df.all$variant_id))
if(varcount)
{
  viewDataTable(dat = var.df.all)
}
```
**There are a total of `r varcount` variants**
</div>
</div>

```{r ar, warnings=FALSE, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
var.df<-execar.df
if(length(var.df) == 1 && var.df == 0)
{
  varcount <- 0
  showChunk <- FALSE
} else {
  showChunk <- TRUE
}
```
```{r var_format_ar, eval=showChunk, ref.label='var_format', echo=FALSE, message=FALSE, cache=FALSE,  fig.path='{{SLINK}}/'}
```

<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Autosomal recessive</h3>
</div>
<div class="panel-body">
Variants matching following gemini query:
```
gemini query -q "{input.gemini_tool} autosomal_recessive -d 15 --min-gq 80  
--filter "chrom NOT LIKE '%GL%' and (exac_af_all = null or
exac_af_all<0.01) and (rvis_pct<90 or rvis_pct='NA' or rvis_pct='None')
and (impact!='synonymous_variant' and impact!='5_prime_UTR_variant' and
impact!='3_prime_UTR_variant' and impact!='intron_variant') --header"
{input.gemini_db} > {output.gemini_out}"
{db}
```

```{r var_ar, eval=showChunk, warning=FALSE, echo=FALSE, message=FALSE, cache=FALSE,  fig.path='{{SLINK}}/'}
var.df.all <- var.df

varcount <- length(unique(var.df.all$variant_id))
if(varcount)
{
  viewDataTable(dat = var.df.all)
}
```
**There are a total of `r varcount` variants**
</div>
</div>


```{r ch, warnings=FALSE, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
var.df<-execch.df
if(length(var.df) == 1 && var.df == 0)
{
  varcount <- 0
  showChunk <- FALSE
} else {
  showChunk <- TRUE
}
```
```{r var_format_ch, eval=showChunk, ref.label='var_format', echo=FALSE, message=FALSE, cache=FALSE,  fig.path='{{SLINK}}/'}
```


<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Compound het</h3>
</div>
<div class="panel-body">
Variants matching following gemini query:
```
{input.gemini_tool} comp_hets -d 15 --min-gq 80 \
--filter "chrom NOT LIKE '%GL%' and (exac_af_all = null or exac_af_all<0.01) \
and (rvis_pct<90 or rvis_pct='NA' or rvis_pct='None') and (impact!='synonymous_variant' \
and impact!='5_prime_UTR_variant' and impact!='3_prime_UTR_variant')" --max-priority 1 \
--header {input.gemini_db} > {output.gemini_out}
```

```{r var_ch, eval=showChunk, warning=FALSE, echo=FALSE, message=FALSE, cache=FALSE,  fig.path='{{SLINK}}/'}
var.df.all <- var.df

varcount <- length(unique(var.df.all$variant_id))
if(varcount)
{
  viewDataTable(dat = var.df.all)
}
```
**There are a total of `r varcount` variants**
</div>
</div>


<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Meta</h3>
</div>
<div style="color:#003366;", class="panel-body">
Git commit: **`r commit`**
</div>
</div>
