---
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    lib_dir: libs
css: styles.css
params: 
    title: "GRIN Epilepsy Trio Analysis!"
---
---
title: `r params$title`
---
```{r libs, echo=FALSE, message=FALSE, cache=FALSE, warning=FALSE, results='hide', fig.path='{{SLINK}}/'}
library(knitr)
library(DT)
library(plyr)
library(dplyr)
library(GenomicRanges)
library(stringr)
library(reshape2)
options(digits = 2)
options(width=1200)
```

```{r commit, echo=FALSE, message=FALSE, cache=FALSE, warning=FALSE, results='hide', fig.path='{{SLINK}}/'}
commit <- system("git rev-parse --verify HEAD",intern=TRUE)
```

## Filters
<blockquote>
We used the following filters to reduce the number of reported false positives:

1. Spurious inherited variants can often appear due to insufficient sequence coverage so the **mininum sequence depth** was set to **15**.
2. We are interested in mutations that are likely to have functional consequence so only variants with **Medium or High impact severity** were used.
3. **Minimum genotyping quality** per sample in the trio was set to **20**.
4. GL contigs were not used.
</blockquote>

## Compound Heterozygotes
<blockquote>
Genotype Requirements:
  
1. All affected individuals must be heterozygous at both sites.
2. No unaffected can be homozygous alterate at either site.
3. Neither parent of an affected sample can be homozygous reference at both sites.
4. If any unphased-unaffected is het at both sites, the site will be give lower priority
5. No phased-unaffected can be heterozygous at both sites.
6. Remove candidates where an affected from the same family does NOT share the same het pair.
7. Sites are automatically phased by transmission when parents are present in order to remove false positive candidates.
</blockquote>

<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Variant summary</h3>
</div>
<div class="panel-body">

This represents a brief summary of variants.

```{r var_summary, results='asis', warning=FALSE, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}

if(length(var.df) == 1 && var.df == 0)
{
  varcount <- 0
  showChunk <- FALSE
} else {
  showChunk <- TRUE
  dt <- plyr::count(var.df , 'impact_so')
  viewDataTable(dat = data.frame(Description = dt$impact_so, 
                                 Variants = dt$freq, 
                                 Perc = round(prop.table(dt$freq)*100,2)))
}
```

```{r var_igv, warnings=FALSE, eval=FALSE, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
var.df$variant_id <- paste("<a href=\"", paste(paste("../screenshots", paste(familyname, model, var.df$variant_id, 'png', sep="."), sep="/")), "\">", var.df$variant_id, "</a>", sep="")
```

```{r var_rvis, warnings=FALSE, eval=showChunk, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
var.df$orig_chrom <- var.df$chrom
var.df$chrom <- sub('^chr','',var.df$chrom)
var.df <- merge(var.df, rvis, by.x = "gene", by.y = "GENE", all.x = T, sort = F)
```

```{r var_formatcols, warnings=FALSE, eval=showChunk, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
var.df$start <- ifelse(var.df$end-var.df$start==1, var.df$end, paste0(var.df$start,'-',var.df$end))
colnames(var.df)[grep('^chrom$|start|vcf_id', colnames(var.df))] <- c('chr','pos','rsid')
var.df$end <- NULL
var.df <- var.df[,c(2:16,1,17:ncol(var.df))]

chr <- c(1:22,'X','Y','MT')
var.df$chr <- factor(var.df$chr, levels = chr)
var.df <- var.df[order(var.df$chr),]
var.df$exac_ac_all[var.df$exac_ac_all=="None"] <- "-1"
var.df$exac_ac_all <- as.numeric(var.df$exac_ac_all)
```

```{r var_filtered_variants, warning=FALSE, eval=showChunk, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
varcount <- length(unique(var.df$variant_id))
```

**There are `r varcount` variants.**
</div>
</div>

<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">All variants</h3>
</div>
<div class="panel-body">
These are all variants

```{r var_all, eval=showChunk, warning=FALSE, echo=FALSE, message=FALSE, cache=FALSE,  fig.path='{{SLINK}}/'}
var.df.all <- var.df

varcount <- length(unique(var.df.all$variant_id))
if(varcount)
{
  viewDataTable(dat = var.df.all)
}
```
**There are a total of `r varcount` variants**
</div>
</div>

<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Variants for Ingo Helbig (Enhanced)</h3>
</div>
<div class="panel-body">
These are all variants that satisfy the following criteria:

1. RVIS < 90%.
2. EXAC frequency <= 1% (or Unknown) only if both variants in the cmpd het are common.
3. Annotated by gene instances frequency.

```{r var_filtered_enhancement, eval=showChunk, warning=FALSE, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
var.df.sub <- var.df
var.df.sub <- var.df.sub %>% filter(rvis_perc < 90) %>% group_by(comp_het_id) %>% filter(any(exac_af_all <= 0.1) | any(exac_af_all=="None")) %>% group_by(gene) %>% mutate(gene_freq = n())
varcount <- length(unique(var.df.sub$variant_id))
if(varcount)
{
  viewDataTable(dat = var.df.sub)
}
```
**There are `r varcount` filtered variants**
</div>
</div>

<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Triage 1: Variants in recurrent genes</h3>
</div>
<div class="panel-body">
These are variants found in the recurrent genes associated with recurrent mutations in epilepsy.

```{r var_recurrent, results='asis', warning=FALSE, eval=showChunk, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
var.df %>% dplyr::filter(recurrent_genes != "None") -> in_recurrent 
varcount <- length(unique(in_recurrent$variant_id))
if(varcount)
{
  viewDataTable(dat = in_recurrent)
}
```
**There are `r varcount` variants in recurrent genes.**

</div>
</div>

<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Triage 2: Variants in literature genes with low tolerance</h3>
</div>
<div class="panel-body">
These are the remaining variants cited in the literature genes cited in the existing literature. Many of the literature-cited genes are likely tolerant, so this list is further filtered to only genes with an RVIS "ALL_0.1." value of <= 0.

```{r var_literature, results='asis', warning=FALSE, eval=showChunk, echo=FALSE, message=FALSE, cache=FALSE, fig.path='{{SLINK}}/'}
var.df %>% dplyr::filter(literature_genes != "None") %>% dplyr::filter(as.numeric(rvis_score)<=0) %>% thisnotinthat(in_recurrent) -> in_literature
varcount <- length(unique(in_literature$variant_id))
if(varcount)
{
  viewDataTable(dat = in_literature)
}
```
**There are `r varcount` remaining variants found in literature genes.**

</div>
</div>

<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Triage 3: Coding variants not in the above</h3>
</div>
<div class="panel-body">
These are remaining variants that are located in the coding sequence of at least one transcript.

```{r var_coding, eval=showChunk, warning=FALSE, echo=FALSE, message=FALSE, cache=FALSE,  fig.path='{{SLINK}}/'}
var.df %>% dplyr::filter(is_coding==1) %>% thisnotinthat(in_literature) -> codingonly
varcount <- length(unique(codingonly$variant_id))
if(varcount)
{
  viewDataTable(dat = codingonly)
}
```
**There are `r varcount` remaining variants found in coding genes**
</div>
</div>

<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">All other variants</h3>
</div>
<div class="panel-body">
These are the remaining variants

```{r var_others, eval=showChunk, warning=FALSE, echo=FALSE, message=FALSE, cache=FALSE,  fig.path='{{SLINK}}/'}
others <- thisnotinthat(var.df, in_literature)
others <- thisnotinthat(others, codingonly)

varcount <- length(unique(others$variant_id))
if(varcount)
{
  viewDataTable(dat = others)
}
```
**There are `r varcount` remaining variants**
</div>
</div>

<div class="panel panel-default">
<div class="panel-heading">
<h3 class="panel-title">Meta</h3>
</div>
<div style="color:#003366;", class="panel-body">
Git commit: **`r commit`**
</div>
</div>